ideas/276-fast-discovery.md

```diff
@@ -0,0 +1,142 @@
+---
+id: 276-fast-discovery
+title: Fast discovery
+status: draft
+created: 2018-06-22
+category: core
+lead-contributor: dshulyak
+contributors:
+        - dshulyak
+exit-criteria: yes
+success-metrics: yes
+clear-roles: yes
+future-iteration: ?
+roles-needed:
+    - Go Developer
+okrs:
+    - "[PO]: Objective: Messaging is reliable"
+    - "[P2]: KR: Cluster can handle 500 concurrent users"
+---
+
+## Preamble
+
+    Idea: #276-fast-discovery
+    Title: Fast discovery
+    Status: draft
+    Created: 2018-06-22
+
+## Summary
+
+Previously we added support for discovery v5 in status client.
+While it allows us to discover peers in under 5 seconds, and consumes on average 50kb of traffic (sum of inbound and outound).
+This picture completely changes with more concurrent clients. If 100 clients joins simultaneously discovery becomes
+very slow and not reliable. For instance it can consume as much as 2mb in 10s while not finding required number of peers.
+
+## Swarm Participants
+
+- Lead Contributor: dshulyak
+
+## Product Overview
+
+We will implement protocol that takes into consideration problems mentioned in summary of this document.
+
+### User Stories
+
+Devices will be able to connect with new peers without spending noticeable amount of time and traffic.
+
+### Requirements & Dependencies
+
+No.
+
+### Security and Privacy Implications
+
+Same properties.
+
+## Protocol implementation.
+
+### Protocol
+
+https://github.com/libp2p/specs/blob/4059338ff0f90835ed953e1eb4c2beb703cc04d9/rendezvous/README.md
+
+For details please see that spec. but essentially it just implements 3 RPC call.
+1. discover peers
+2. register peer for a particular topic
+3. unregister peer
+
+### Peer definition
+
+We need to share more information than pubkey of the peer and information
+for ip connection. There is a format suitable for sharing any
+key=value information. You can learn more about ENRs here:
+
+https://github.com/ethereum/EIPs/blob/master/EIPS/eip-778.md
+
+One of the goals is to advertise provider signature.
+For details see discoverable mail server swarm.
+
+### Transport layer
+
+We are free in the choice of the transport. I suggest to use go-stream-muxer
+for negotiating connection type (https://github.com/libp2p/go-stream-muxer).
+With a preference towards rlp stream over tcp connection.
```

Comment:
Why would rlp stream be preferred? The spec provides a protobuf file unless we don't want to introduce protobuf dependency which is quite large as far as I remember.

Reply:
it can be protobuf too, but we wont be compatible with original spec because of enr usage anyway. so i decided that we may as well dont use protobuf. and use rlp, cause there is already couple of
helpers for marshaling/unmarshaling ENR with rlp

Reply:
That sounds good. RLP then 👍